provide a different mechanism for metrics collection

bobby-stripe opened the pull request:
Today, we provide a way for callers to pass in functions that get deferred and called when a flag is evaluated so that we can track FF accesses.  The straightforward way for callers to do this, however, involves having a big `sync.Map` which contents.

This PR instead adds 2 `sync.Uint64`s per feature flag, and lets a caller periodically call `ReportCounts` to read (and reset, for the next interval) the counts per flag.  This avoids contention between flags, and is very fast.


It also lurks in removing the unused CSV backend.  I had hoped to get rid of the `Flags1` side of things altogether, but its still used by at least 1 internal user.

vasi-stripe commented on goforit.go:
```diff
@@ -206,9 +206,12 @@ func DeletedCallback(cb evaluationCallback) Option {
+	disabledCount atomic.Uint64
+	enabledCount  atomic.Uint64
+	clamp         clamp.Clamp
+	// pad out to end of 64-byte cacheline to prevent false sharing
```
😬

bobby-stripe replied:
@vasi-stripe I revised this, and the struct is now only 32-bytes wide (with no padding) -- half the size of a cacheline on intel/amd/graviton hosts.  This ensures that it doesn't "straddle" a cacheline, which is good enough!

vasi-stripe commented:
Let's also remove the CSV files in testdata

vasi-stripe commented on goforit.go:
```diff
@@ -424,5 +441,20 @@ func (g *goforit) Close() error {
+	g.reportMu.Lock()
+	defer g.reportMu.Unlock()
+
+	for name, fh := range g.flags.load() {
```
We can have rather a lot of flags to iterate over! I guess it's still preferable to move this out of the synchronous Enabled calls, but it could be a lot of overhead for services that use flags rarely.

bobby-stripe replied:
@vasi-stripe yeah - I think it should be fine, as we're walking a map without a lock that affects Enabled held, but I added a metric here so that we can see how long this actually takes in QA and prod in a few specific canary services.

vasi-stripe commented:
r? @bobby-stripe

bobby-stripe commented:
r? @vasi-stripe 

a few things:
* probably worth reviewing https://github.com/stripe/goforit/pull/51 first, as that is _just_ the removal of the v1 flags backends
* got rid of the explicit cacheline padding
* added tests!
* added metrics to ensure the new Report interface isn't too slow with a lot of flags.

vasi-stripe commented on goforit.go:
```diff
@@ -2,6 +2,7 @@ package goforit
 
 import (
 	"context"
+	"github.com/stripe/goforit/flags2"
```
Needs grouping

vasi-stripe reviewed (changes requested):
Just a couple of nits

bobby-stripe commented:
r? @vasi-stripe 

thanks, good catches!  I rebased + pushed a commit with the fixes

bobby-stripe commented:
sorry, need 1 more boop - I made a tiny change to also report back if the flag is deleted
